### PR TITLE
Setup goreleaser and improve CI

### PR DESCRIPTION
**Pull request by rvolosatovs**
**Summary:**

References #86 

**Changes:**

- Remove `ci` and go build-related `Makefile`/`mage` logic
- Simpify `.travis.yml` and the matrix
- Setup `goreleaser`
- Vendor tools via Go modules(see https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module) - no more `go.dev-deps`

**Notes for Reviewers:**

This is not a "final" version of `.goreleaser.yml`, more like a "base" to work with.
I suppose we will gather feedback from actual users of the packages(maybe @proffalken ?) and improve them based on that. For starters, we probably need to create a wrapper for `ttn-lw-stack` to specify the actual frontend location, same as done for Brew. I will create an issue about it.(TODO: add ref)
This is what `goreleaser` does for now(the repositories in the PR are in `TheThingsNetwork` namespace `rvolosatovs` was used for testing):
- Creates a release with changelog and references to Docker images: https://github.com/rvolosatovs/lorawan-stack/releases/tag/v3.0.1
- Builds docker images and pushes to Docker Hub: https://cloud.docker.com/u/rvolosatovs/repository/docker/rvolosatovs/lorawan-stack
- Creates a brew tap formula and pushes to tap repository:
https://github.com/rvolosatovs/homebrew-lorawan-stack
You can try the Brew package by `brew install rvolosatovs/lorawan-stack/ttn-lw-stack`(it's going to be `brew install TheThingsNetwork/lorawan-stack/ttn-lw-stack` once we merge this and create a tag)
(Thanks @bafonins for testing)

We don't need any Go build logic in `make` or `mage` anymore, users should be able to just `go get` or `go build` as needed(cc @htdvisser ). The binary releases are built by goreleaser. For testing the "full package" can be built by `goreleaser --skip-publish --snapshot` locally if required.

**Release Notes: _(optional)_**

- LoRaWAN stack and CLI are now available via `brew`: just do `brew install TheThingsNetwork/lorawan-stack/ttn-lw-stack`

### CODE REVIEW

File: .make/go/test.make [hunk 1]
```diff
@@ -14,8 +14,8 @@
 
 # run tests
 go.test: go.min-version key.pem cert.pem
-	@$(log) "Testing `$(TEST_PACKAGES) | $(count)` go packages"
-	$(GO) test $(GO_FLAGS) $(GO_TEST_FLAGS) `$(TEST_PACKAGES)`
+	@$(log) "Running tests..."
+	$(GO) test $(GO_FLAGS) $(GO_TEST_FLAGS) ./...
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply .make/go/test.make]: https://golang.org/doc/go1.9#vendor-dotdotdot
We require `1.11` for our builds anyway, so no reason be compatible with pre-`1.9`
Besides, by using `./...` we also ensure that packages, which do not contain tests and are not imported by tested packages actually compile(e.g. `cmd`) and since we don't do `build-all` on PRs anymore, we need to test that e.g. CLI can be built.

File: .mage/go.go [hunk 2]
```diff
@@ -106,7 +106,7 @@ var goBuildConfigs = []goBuildConfig{
 	{GOOS: "darwin", GOARCH: "amd64"},
 }
 
-var releaseDir = "release"
+var releaseDir = "dist"
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply .mage/go.go]: `goreleaser` uses `dist` by default, it makes sense to be compatible. See https://goreleaser.com/environment/
We don't need this file at all anymore, the only reason to keep it around is to be able to do convenient cross-platform builds without using `goreleaser`.
`goreleaser` is not installed as part of `dev-tools`, but only used in CI, so I kept this here.
What we can do instead is to just install `goreleaser` and do `goreleaser --skip-publish` to build all the binaries etc.
Shall I just remove this file and add `goreleaser` to `dev-deps`?

File: .goreleaser.yml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        - '^ci:'
+        - '^Merge pull request'
+
+release:
+  prerelease: auto
+
+nfpm:
+  vendor: The Things Network
+  homepage: https://thethingsnetwork.org
+  maintainer: Roman Volosatovs <roman@thethingsnetwork.org>
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply .goreleaser.yml]: Is this forwarded to all of us or how does it work? I did not even know such an email exists.

File: .mage/go.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply .mage/go.go]: Removed the build logic and made `Makefile` just use vendored `goreleaser` for (test) builds.

File: .goreleaser.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (johanstokking) [reply .goreleaser.yml]: Make it `stack@` and I'll make sure it's being forwarded accordingly.

[the remark this answers is not in the document]
Comment (johanstokking) [reply .goreleaser.yml]: For the record; this is done now.

File: .mage/go.go [hunk 4]
```diff
@@ -53,135 +49,3 @@ func (Go) CheckVersion() error {
...
-		return "", err
-	}
-	var flags []string
-	for k, v := range map[string]string{
-		"GitCommit": commit,
-		"GitBranch": branch,
-		"TTN":       tag,
-		"BuildDate": now.Format("2006-01-02T15:04:05Z"),
-	} {
-		flags = append(flags, fmt.Sprintf("-X %s/pkg/version.%s=%s", goModule, k, v))
```
Comment (htdvisser) [inline .mage/go.go]: How do we insert the build/version info now?

File: .goreleaser.yml [hunk 5]
```diff
@@ -0,0 +1,116 @@
+builds:
+  - main: ./cmd/ttn-lw-stack
+    binary: ttn-lw-stack
```
Comment (htdvisser) [inline .goreleaser.yml]: If this is built into the workdir (lorawan-stack) it should be added to gitignore. Same for cli below.

File: .mage/go.go (hunk 4 above)
Comment (rvolosatovs) [reply .mage/go.go]: See https://goreleaser.com/build/
The default `ldflags` are `-s -w -X main.version={{.Version}} -X main.commit={{.ShortCommit}} -X main.date={{.Date}}`
Isn't that enough?(also in response to https://github.com/TheThingsNetwork/lorawan-stack/pull/109#pullrequestreview-201993397)

Comment (htdvisser) [reply .mage/go.go]: OK with me, but then `main` needs to set those in `pkg/version`, as those are the ones that are actually used by the stack.

File: .goreleaser.yml (hunk 5 above)
Comment (rvolosatovs) [reply .goreleaser.yml]: It's not - `goreleaser` creates a `dist` directory, where all the artifacts are placed.

Comment (htdvisser) [reply .goreleaser.yml]: Then we probably need to check the Dockerfile, because https://github.com/TheThingsNetwork/lorawan-stack/pull/109/files#diff-3254677a7917c6c01f55212f86c57fbfR7 copies it from the workdir. We should also probably document how `docker build` needs to be executed since you removed the `make docker` target.

File: Makefile [hunk 6]
```diff
@@ -29,44 +29,25 @@ include .make/go/main.make
...
-dev-deps: go.dev-deps js.dev-deps
+dev-deps: go.deps js.dev-deps
 
 deps: go.deps js.deps sdk.deps
 
 test: go.test js.test sdk.test
 
 quality: go.quality js.quality styl.quality snap.quality
 
-clean: go.clean js.clean
```
Comment (htdvisser) [inline Makefile]: Why remove `make clean`?

File: .goreleaser.yml (hunk 5 above)
Comment (rvolosatovs) [reply .goreleaser.yml]: Haha, I initially wrote this in https://github.com/TheThingsNetwork/lorawan-stack/pull/109#discussion_r255435612 but then deleted

For `docker`, `goreleaser` creates a temporary directory and copies the required binaries and any `extra_files` specified to the root of that temporary directory.
You can verify that binaries are there by looking/testing `rvolosatovs/lorawan-stack` Docker image(https://cloud.docker.com/u/rvolosatovs/repository/docker/rvolosatovs/lorawan-stack)

File: Makefile (hunk 6 above)
Comment (rvolosatovs) [reply Makefile]: It's still here
https://github.com/TheThingsNetwork/lorawan-stack/pull/109/files/dd77b649359aa917a56980930e0a6056355e8381#diff-b67911656ef5d18c4ae36cb6741b7965R50

File: .mage/go.go (hunk 4 above)
Comment (rvolosatovs) [reply .mage/go.go]: This is what we get now(https://github.com/rvolosatovs/lorawan-stack/releases/tag/v3.0.2):
```
docker run --rm rvolosatovs/lorawan-stack:3.0.2 cli version
The Things Network Stack for LoRaWAN: ttn-lw-cli
Version:             3.0.2
Go version:          go1.11.5
OS/Arch:             linux/amd64
```

File: .goreleaser.yml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  hooks:
+  - make js.build
+
+builds:
+  - main: ./cmd/ttn-lw-stack
+    binary: ttn-lw-stack
+    ldflags:
+    - -s
+    - -w
+    - -X go.thethings.network/{{.ProjectName}}/pkg/version.BuildDate={{.Date}}
```
Comment (htdvisser) [inline .goreleaser.yml]: I don't think it's appropriate to use `{{.ProjectName}}` here. The go module path is `go.thethings.network/lorawan-stack/pkg/version`, which isn't necessarily the same as the name of the dir in which this repo is cloned.

File: .goreleaser.yml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    (bin/"ttn-lw-stack").write_env_script libexec/"ttn-lw-stack", env
+
+dockers:
+  - goos: linux
+    goarch: amd64
+    binaries:
+      - ttn-lw-cli
+      - ttn-lw-stack
+    image_templates:
+      - "TheThingsNetwork/lorawan-stack:latest"
```
Comment (htdvisser) [inline .goreleaser.yml]: Docker images are lowercase, use `thethingsnetwork/lorawan-stack`,

File: .goreleaser.yml (hunk 7 above)
Comment (rvolosatovs) [reply .goreleaser.yml]: I'll set `project_name` explicitly then, since this may affect other artifacts as well

Comment (htdvisser) [reply .goreleaser.yml]: Fine, but that doesn't change anything about the fact that the go module path is something different.

Comment (rvolosatovs) [reply .goreleaser.yml]: How is that related here?

Comment (htdvisser) [reply .goreleaser.yml]: `go.thethings.network/{{.ProjectName}}` in these lines must change to `go.thethings.network/lorawan-stack`.

File: DEVELOPMENT.md [hunk 9]
```diff
@@ -77,15 +77,24 @@ make test
 
 ### Building
 
-You can build the stack with the following command:
+There is a single binary for the server, `ttn-lw-stack`, as well as a binary for the command-line interface `ttn-lw-cli`. The single binary contains all components start one or multiple components. This allows you to run the stack with one command in simple deployment scenarios, as well as distributing micro-services for more advanced scenarios.
+
+We provide binary releases for all supported platforms, including packages for various package managers at https://github.com/TheThingsNetwork/lorawan-stack/releases . We suggest you use the compiled packages we provide in production scenarios.
+
+For development/testing purposes we suggest either running required binaries via `go run`(e.g. `go run ./cmd/ttn-lw-cli` from repository root for CLI), or using `go build` directly.
```
Comment (htdvisser) [inline DEVELOPMENT.md]: Now that we no longer recommend `make clean build-all` we should also give instructions on how to build the frontend during development, right? CC: @bafonins @kschiffer 

Can be a separate PR as far as I'm concerned.
